### PR TITLE
Expose metrics via Control

### PR DESCRIPTION
**Pull request by ktoso**
We'd like to be able to expose metrics to users when they use the Sharing the Consumer Actor pattern; I thought this style would likely be good enough - I assume the same could be done in the Publisher?

### CODE REVIEW

**Comment by ktoso** [comment]
Please review is this general idea could make it in @patriknw @kciesielski @13h3r :)

**Comment by 13h3r** [comment]
Yes, I think it could be implemented this way

**Comment by kciesielski** [comment]
Thanks, LGTM. As for producer, there's `Producer.flow(settings: ProducerSettings, producer: KafkaProducer)`, so users can provide their own producers which can be directly asked for metrics.

**Comment by ktoso** [comment]
thanks, will finish this PR Then soon

On Sep 10, 2017 11:47 AM, "Krzysiek Ciesielski" <notifications@github.com>
wrote:

Thanks, LGTM. As for producer, there's Producer.flow(settings:
ProducerSettings, producer: KafkaProducer), so users can provide their own
producers which can be directly asked for metrics.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/akka/reactive-kafka/pull/338#issuecomment-328331691>,
or mute the thread

File: core/src/main/scala/akka/kafka/KafkaConsumerActor.scala [hunk 1]
```diff
@@ -48,7 +52,17 @@ object KafkaConsumerActor {
 
     private[KafkaConsumerActor] class NoPollResult extends RuntimeException with NoStackTrace
   }
-
+  
+  /** Public protocol which can be used to interact with the Actor */
+  object Protocol {
+    // requests
+    final case class RequestMetrics()
```
Comment (patriknw) [inline core/src/main/scala/akka/kafka/KafkaConsumerActor.scala]: strange with case class without params, case object?

then it need accessor for Java API

File: core/src/main/scala/akka/kafka/KafkaConsumerActor.scala [hunk 2]
```diff
@@ -48,7 +52,17 @@ object KafkaConsumerActor {
...
+    
+    // responses
+    /** Contains a snapshot of metrics obtained from the underlying kafka consumer */
+    final case class ConsumerMetrics(metrics: Map[MetricName, Metric])
```
Comment (patriknw) [inline core/src/main/scala/akka/kafka/KafkaConsumerActor.scala]: Java API: getMetrics: java.util.Map

**Comment by v-gerasimov** [comment]
What about this PR? Seems there is no way to get consumer metrics now.

**Comment by ktoso** [comment]
Hah, got reminded of this one via discuss.akka.io Fill fix it

File: core/src/main/scala/akka/kafka/KafkaConsumerActor.scala (hunk 1 above)
Comment (ktoso) [reply core/src/main/scala/akka/kafka/KafkaConsumerActor.scala]: ok

**Comment by ktoso** [comment]
Addressed feedback, checked Java API, added docs (also for producer where it's trivial however added it so it's all explained symmetrically).

Please review @kciesielski @patriknw

File: core/src/main/scala/akka/kafka/KafkaConsumerActor.scala [hunk 3]
```diff
@@ -61,6 +77,15 @@ object KafkaConsumerActor {
...
   }
 
+  private[kafka] def rebalanceListener(onAssign: Iterable[TopicPartition] => Unit, onRevoke: Iterable[TopicPartition] => Unit) = new ConsumerRebalanceListener {
+    override def onPartitionsAssigned(partitions: util.Collection[TopicPartition]): Unit = {
+      onAssign(partitions.asScala)
+    }
+    override def onPartitionsRevoked(partitions: util.Collection[TopicPartition]): Unit = {
+      onRevoke(partitions.asScala)
+    }
+  }
```
Comment (patriknw) [inline core/src/main/scala/akka/kafka/KafkaConsumerActor.scala]: what does this have to do with metrics?

File: core/src/main/scala/akka/kafka/KafkaConsumerActor.scala [hunk 4]
```diff
@@ -11,25 +11,41 @@ import java.util.concurrent.TimeUnit
...
+  case object RequestMetrics extends RequestMetrics {
+    /** Java DSL */
+    def getInstance: RequestMetrics = this
+  }
+  /** Contains a snapshot of metrics obtained from the underlying kafka consumer */
+  final case class ConsumerMetrics(metrics: Map[MetricName, Metric]) {
+    /** Java API */
+    def getMetrics: util.Map[MetricName, Metric] = metrics.asJava
+  }
+  // ------------ end of public protocol ------------
```
Comment (patriknw) [inline core/src/main/scala/akka/kafka/KafkaConsumerActor.scala]: This works when the KafkaConsumer actor is created and passed in, but what about the other cases where the actor is created by the library? Wouldn't it be better to expose the metrics via the `Control` materialized value?

Could be a Future and using ask internally.

File: core/src/main/scala/akka/kafka/KafkaConsumerActor.scala (hunk 3 above)
Comment (ktoso) [reply core/src/main/scala/akka/kafka/KafkaConsumerActor.scala]: Rebase goone weird :\ Thanks

**Comment by ktoso** [comment]
Yeah you're right that a `def metrics` on the `Control` should be added as well; will do so

File: core/src/main/scala/akka/kafka/KafkaConsumerActor.scala (hunk 4 above)
Comment (ktoso) [reply core/src/main/scala/akka/kafka/KafkaConsumerActor.scala]: Yeah good point, will add that

**Comment by patriknw** [comment]
Perhaps `Control` can be the only public API?

**Comment by ktoso** [comment]
> Perhaps Control can be the only public API?

Yeah, tried doing that. It's hard to make it `def metrics: Map[MetricName, Metric]` sadly, due to the SubSourceLogic and friends... This would be optimal API for users.

It is easier to make it a `Future[Map` which probably is good enough

**Comment by ktoso** [comment]
Done, now the only API is to get metrics from the control value

**Comment by ktoso** [comment]
Please give it a look @patriknw when you have some time

File: core/src/main/scala/akka/kafka/internal/SingleSourceLogic.scala [hunk 5]
```diff
@@ -6,16 +6,21 @@
...
+import akka.kafka.KafkaConsumerActor.Internal.{ConsumerMetrics, RequestMetrics}
 import akka.kafka.Subscriptions._
 import akka.kafka.{ConsumerFailed, ConsumerSettings, KafkaConsumerActor, Subscription}
 import akka.stream.stage.GraphStageLogic.StageActor
 import akka.stream.stage.{GraphStageLogic, OutHandler, StageLogging}
 import akka.stream.{ActorMaterializerHelper, SourceShape}
-import org.apache.kafka.clients.consumer.ConsumerRecord
-import org.apache.kafka.common.TopicPartition
+import akka.util.Timeout
+import org.apache.kafka.clients.consumer.{Consumer, ConsumerRecord}
```
Comment (chbatey) [inline core/src/main/scala/akka/kafka/internal/SingleSourceLogic.scala]: Consumer is unused

File: core/src/main/scala/akka/kafka/internal/SubSourceLogic.scala [hunk 6]
```diff
@@ -9,6 +9,8 @@ import java.util.concurrent.TimeUnit
 
 import akka.NotUsed
 import akka.actor.{ActorRef, ExtendedActorSystem, Terminated}
+import akka.dispatch.ExecutionContexts
```
Comment (chbatey) [inline core/src/main/scala/akka/kafka/internal/SubSourceLogic.scala]: Clean up imports, some of the new unused

File: docs/src/test/scala/sample/scaladsl/ConsumerExample.scala [hunk 7]
```diff
@@ -7,20 +7,23 @@ package sample.scaladsl
 import akka.pattern.ask
 import akka.kafka.ConsumerMessage.{CommittableMessage, CommittableOffsetBatch}
 import akka.kafka._
-import akka.actor.{Props, ActorRef, Actor, ActorSystem, ActorLogging, PoisonPill}
+import akka.actor.{Actor, ActorLogging, ActorRef, ActorSystem, PoisonPill, Props}
```
Comment (chbatey) [inline docs/src/test/scala/sample/scaladsl/ConsumerExample.scala]: Some of these new imports are unused

File: core/src/main/scala/akka/kafka/KafkaConsumerActor.scala [hunk 8]
```diff
@@ -187,6 +192,11 @@ private[kafka] class KafkaConsumerActor[K, V](settings: ConsumerSettings[K, V])
         stopInProgress = true
         context.become(stopping)
       }
+
+    case RequestMetrics =>
+      val unmodifiableYetMutableMetrics: java.util.Map[MetricName, _ <: Metric] = consumer.metrics()
```
Comment (chbatey) [inline core/src/main/scala/akka/kafka/KafkaConsumerActor.scala]: This map goes through a lot of conversions for the Java case.

I guess there's no point trying to avoid this. Looking at the Kafka code they give us the raw map wrapped in an unmodifiable so we shouldn't pass that around.

Comment (ktoso) [reply core/src/main/scala/akka/kafka/KafkaConsumerActor.scala]: Yeah it is converted a lot indeed... I figured to leave it along until reported as a problem

**Comment by ktoso** [comment]
Thanks! Exciting one :)
